Source/WebCore/page/writing-tools/WritingToolsController.mm

```diff
@@ -929,6 +929,119 @@ static bool appliedCommandIsWritingToolsCommand(const Vector<Ref<WritingToolsCom
+        auto data = std::get<DocumentMarker::WritingToolsTextSuggestionData>(marker.data());
+
+        auto markerRange = makeSimpleRange(node, marker);
+
+        auto rect = document->view()->contentsToRootView(unionRect(RenderObject::absoluteTextRects(markerRange, { })));
+        textRectsInRootViewCoordinates.append(WTFMove(rect));
+
+        return false;
+    });
+
+    return textRectsInRootViewCoordinates;
+}
+
+void WritingToolsController::updateTextVisibilityForActiveSession(const CharacterRange& rangeRelativeToSessionRange, bool visible)
+{
+    RefPtr document = this->document();
+    if (!document) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    auto sessionRange = this->activeSessionRange();
+    if (!sessionRange) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    auto resolvedRange = resolveCharacterRange(*sessionRange, rangeRelativeToSessionRange);
+
+    if (visible)
+        document->markers().removeMarkers(resolvedRange, { WebCore::DocumentMarker::Type::TransparentContent });
+    else {
+        // FIXME: Remove the UUID parameter once the old animation system is removed and it's no longer needed.
+        document->markers().addTransparentContentMarker(resolvedRange, WTF::UUID { 0 });
+    }
+}
+
+std::optional<TextIndicatorData> WritingToolsController::textPreviewDataForActiveSession(const CharacterRange& rangeRelativeToSessionRange)
+{
+    RefPtr document = this->document();
+    if (!document) {
+        ASSERT_NOT_REACHED();
+        return std::nullopt;
+    }
+
+    auto sessionRange = this->activeSessionRange();
+    if (!sessionRange) {
+        ASSERT_NOT_REACHED();
+        return std::nullopt;
+    }
+
+    auto resolvedRange = resolveCharacterRange(*sessionRange, rangeRelativeToSessionRange);
+
+    // Temporarily remove any transparent content document markers so that when the snapshot is created, the text is visible.
+    // The markers are then re-added in the same run loop, so there will be no user-visible flickering of the text.
+
+    auto& markers = document->markers();
+
+    Vector<SimpleRange> markerRanges;
+
+    markers.forEach(resolvedRange, { WebCore::DocumentMarker::Type::TransparentContent }, [&](auto& node, auto& marker) {
+        auto markerRange = makeSimpleRange(node, marker);
+        markerRanges.append(markerRange);
+
+        return false;
+    });
+
+    static constexpr OptionSet textIndicatorOptions {
+        TextIndicatorOption::IncludeSnapshotOfAllVisibleContentWithoutSelection,
+        TextIndicatorOption::ExpandClipBeyondVisibleRect,
+        TextIndicatorOption::SkipReplacedContent,
+        TextIndicatorOption::RespectTextColor,
+    };
+
+    RefPtr textIndicator = WebCore::TextIndicator::createWithRange(resolvedRange, textIndicatorOptions, WebCore::TextIndicatorPresentationTransition::None, { });
+    if (!textIndicator)
+        return std::nullopt;
+
+    // Add back any of the transparent content markers which were removed.
```

Comment:
Nit - I think this comment could be removed if you renamed `markerRanges` to something like `transparentMarkerRangesToReinsert`